test/envtest/gatewayclass_envtest_test.go

```diff
@@ -0,0 +1,302 @@
+			GatewayClass: gatewayv1beta1.GatewayClass{
+				Spec: gatewayv1beta1.GatewayClassSpec{
+					ControllerName: unsupportedControllerName,
+				},
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "unsupported-gateway-class",
+				},
+			},
+			Gateway: gatewayv1beta1.Gateway{
+				Spec: gatewayv1beta1.GatewaySpec{
+					GatewayClassName: gatewayv1beta1.ObjectName("unsupported-gateway-class"),
+					Listeners: builder.NewListener("http").
+						HTTP().
+						WithPort(80).
+						WithAllowedRoutes(builder.NewAllowedRoutesFromAllNamespaces()).
+						IntoSlice(),
+				},
+				ObjectMeta: metav1.ObjectMeta{
+					Name: uuid.NewString(),
+				},
+			},
+			Test: func(ctx context.Context, t *testing.T, gwClient gatewayclientv1beta1.GatewayInterface, gwc gatewayv1beta1.GatewayClass, gw gatewayv1beta1.Gateway) {
+				t.Helper()
+
+				t.Logf("deploying gateway class %s", gwc.Name)
+				require.NoError(t, client.Create(ctx, &gwc))
+				t.Cleanup(func() { _ = client.Delete(ctx, &gwc) })
+
+				t.Logf("verifying that the unsupported Gateway %s does not get Accepted or Programmed by the controller", gw.Name)
+				for i := 0; i < 100; i++ {
+					gateway, err := gwClient.Get(ctx, gw.Name, metav1.GetOptions{})
+					require.NoError(t, err)
+					require.Len(t, gateway.Status.Conditions, 2)
+					if lo.ContainsBy(gateway.Status.Conditions, func(c metav1.Condition) bool {
+						return c.Type == "Programmed" && c.Status != metav1.ConditionUnknown
+					}) {
+						t.Fatalf("expected not to find a Programmed condition with Reason different than Unknown, got %#v", gateway.Status.Conditions)
+					}
+					if lo.ContainsBy(gateway.Status.Conditions, func(c metav1.Condition) bool {
+						return c.Type == "Accepted" && c.Status != metav1.ConditionUnknown
+					}) {
+						t.Fatalf("expected not to find a Accepted condition with Reason different than Unknown, got %#v", gateway.Status.Conditions)
+					}
+				}
+			},
+		},
+		{
+			Name: "managed gateway class",
+			GatewayClass: gatewayv1beta1.GatewayClass{
+				Spec: gatewayv1beta1.GatewayClassSpec{
+					ControllerName: gateway.GetControllerName(),
+				},
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "managed-gateway-class",
+				},
+			},
+			Gateway: gatewayv1beta1.Gateway{
+				Spec: gatewayv1beta1.GatewaySpec{
+					GatewayClassName: gatewayv1beta1.ObjectName("managed-gateway-class"),
+					Listeners: builder.NewListener("http").
+						HTTP().
+						WithPort(80).
+						WithAllowedRoutes(builder.NewAllowedRoutesFromAllNamespaces()).
+						IntoSlice(),
+				},
+				ObjectMeta: metav1.ObjectMeta{
+					Name: uuid.NewString(),
+				},
+			},
+			Test: func(ctx context.Context, t *testing.T, gwClient gatewayclientv1beta1.GatewayInterface, gwc gatewayv1beta1.GatewayClass, gw gatewayv1beta1.Gateway) {
+				t.Helper()
+
+				t.Logf("verifying that the Gateway %s does not get scheduled by the controller due to missing its GatewayClass", gw.Name)
+				for i := 0; i < 100; i++ {
```

Comment:
Do we have any other, more deterministic way to tell whether some conditions haven't been set by the controller than checking it arbitrarily 100 times? Maybe it would be possible to decorate the reconcilers' `Reconcile` method to mark it was called, something like:

```go
func (d *Decorator) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
  r, err := d.delegate.Reconcile(ctx, req)
  d.wasCalled = true
  return r, err
}

func (d *Decorator) WasCalled() bool {...}
```

This way we could simply wait for `WasCalled` to become true and then require Conditions to be in the expected state, WDYT?

Reply:
This is unfortunately not possible today and I needed to spent a good hour debugging why 😭 

Basically the TL;DR is that our reconcilers pass themselves to `controller.New()` through `controller.Options` which makes it impossible to wrap it as you suggested above.

Created #4190 to track this and I suggest we leave this as is with the note linking to this issue. WDYT?